src/agent/onefuzz-agent/src/local/tui.rs

```diff
@@ -130,40 +150,97 @@ impl TerminalUi {
+            initial_state.cancellation_tx.subscribe(),
+        ));
 
         let keyboard_ui_event_tx = self.ui_event_tx.clone();
-        let _keyboard_event_handle = Self::read_keyboard_events(keyboard_ui_event_tx);
+        let _keyboard_event_handle = Self::read_keyboard_events(
+            keyboard_ui_event_tx,
+            initial_state.cancellation_tx.subscribe(),
+        );
 
         let task_event_receiver = self.task_event_receiver;
         let ui_event_tx = self.ui_event_tx.clone();
-        let external_event_handle =
-            tokio::spawn(Self::read_commands(ui_event_tx, task_event_receiver));
 
-        let ui_loop = tokio::spawn(Self::ui_loop(initial_state, self.ui_event_rx));
+        let external_event_handle = tokio::spawn(Self::read_commands(
+            ui_event_tx,
+            task_event_receiver,
+            initial_state.cancellation_tx.subscribe(),
+        ));
+
+        let mut task_handles = vec![tick_event_handle, external_event_handle];
+
+        let ui_event_tx = self.ui_event_tx.clone();
+        let telemetry = tokio::spawn(Self::listen_telemetry_event(
+            ui_event_tx,
+            initial_state.cancellation_tx.subscribe(),
+        ));
+
+        task_handles.push(telemetry);
+
+        let ui_loop = tokio::spawn(Self::ui_loop(
+            initial_state,
+            self.ui_event_rx,
+            self.ui_event_tx.clone(),
+        ));
 
-        let mut task_handles = vec![tick_event_handle, ui_loop, external_event_handle];
+        task_handles.push(ui_loop);
 
         if let Some(timeout) = timeout {
             let ui_event_tx = self.ui_event_tx.clone();
-            let timeout_task = tokio::spawn(async move {
-                time::delay_for(timeout).await;
+            tokio::spawn(async move {
+                tokio::time::delay_for(timeout).await;
                 let _ = ui_event_tx.send(TerminalEvent::Quit);
-                Ok(())
             });
-            task_handles.push(timeout_task);
         }
 
-        try_wait_all_join_handles(task_handles)
-            .await
-            .context("ui_loop")?;
+        try_wait_all_join_handles(task_handles).await?;
+        Ok(())
+    }
+
+    async fn listen_telemetry_event(
+        ui_event_tx: UnboundedSender<TerminalEvent>,
+        mut cancellation_rx: broadcast::Receiver<()>,
+    ) -> Result<()> {
+        let mut rx = onefuzz_telemetry::subscribe_to_events();
+
+        while cancellation_rx.try_recv() == Err(broadcast::TryRecvError::Empty) {
+            match rx.recv().await {
+                Ok((event, data)) => {
+                    if let onefuzz_telemetry::Event::coverage_data = event {
+                        let (covered, features, rate) =
+                            data.iter()
+                                .cloned()
+                                .fold((None, None, None), |(c, f, r), d| match d {
+                                    EventData::Covered(value) => (Some(value), f, r),
```

Comment:
Can we include more telemetry data in what's currently listed as coverage?  Would it be difficult to make it such that it was easy to expand the available stats?

For example, when we're fuzzing, we know things like CPU Usage, Physical & Virtual Memory usage, execs/second, etc.

Reply:
@bmc-msft  what do you thing of this 
![image](https://user-images.githubusercontent.com/244326/113951955-84f80700-97c9-11eb-86d6-d639f3782561.png)